dev/io.openliberty.org.jboss.resteasy.common/src/org/jboss/resteasy/core/ServerResponseWriter.java

```diff
@@ -0,0 +1,749 @@
+         boolean hasStarStar = false;
+         boolean hasApplicationStar = false;
+         boolean pFound = false;
+         for (MediaType accept : accepts) {
+            //Instead of getting all the MBWs compatible with type/generic and then filtering using accept, we use
+            //getPossibleMessageBodyWritersMap to get the viable MBWs for the given type AND accept.
+            Map<MessageBodyWriter<?>, Class<?>> mbws = providerFactory.getPossibleMessageBodyWritersMap(type, generic, annotations, accept);
+            for (Entry<MessageBodyWriter<?>, Class<?>> e : mbws.entrySet())
+            {
+               MessageBodyWriter<?> mbw = e.getKey();
+               Class<?> wt = e.getValue();
+               Produces produces = mbw.getClass().getAnnotation(Produces.class);
+               if (produces == null) produces = WILDCARD_PRODUCES; // Spec, section 4.2.3 "Declaring Media Type Capabilities"
+               for (String produceValue : produces.value())
+               {
+                  pFound = true;
+                  MediaType produce = MediaType.valueOf(produceValue);
+
+                  if (produce.isCompatible(accept))
+                  {
+                     SortableMediaType ms = mostSpecific(produce, wt, accept, null);
+                     if (ms.isWildcardSubtype())
+                     {
+                        hasStarStar |= ms.isWildcardType();
+                        hasApplicationStar |= ms.getType().equals("application");
+                     }
+                     M.add(ms);
+                  }
+               }
+            }
+         }
+         if (!pFound) // JAX-RS 2.0 Section 3.8.3
+         {
+            for (MediaType accept : accepts)
+            {
+               MediaType produce = MediaType.WILDCARD_TYPE;
+               if (produce.isCompatible(accept))
+               {
+                  SortableMediaType ms = mostSpecific(produce, null, accept, null);
+                  if (ms.isWildcardSubtype())
+                  {
+                     hasStarStar |= ms.isWildcardType();
+                     hasApplicationStar |= ms.getType().equals("application");
+                  }
+                  M.add(ms);
+               }
+            }
+         }
+         chosen = chooseFromM(chosen, M, hasStarStar, hasApplicationStar);
+      }
+      
+      // Liberty change begins
+      if(chosen.getParameters().containsKey(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES) ||
+                      chosen.getParameters().containsKey("q") || chosen.getParameters().containsKey("qs")) {
+          Map<String, String> map = new HashMap<String, String>(chosen.getParameters());
+          map.remove(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES);
+          map.remove(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES_LC);
+          map.remove("q");
+          map.remove("qs");
+          chosen = new MediaType(chosen.getType(), chosen.getSubtype(), map);
+      }
```

Comment:
```suggestion
      Map<String, String> params = chosen.getParameters();
      int paramsSize = params.size();

      // If there are no parameters, just return the existing chosen 
      if (paramsSize > 0) {
          boolean containsProduces = false;
          boolean containsProducesLC = false;
          boolean containsQ = false;
          boolean containsQS = false;
          int foundParams = 0;
          containsProduces = params.containsKey(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES);
          if (containsProduces = params.containsKey(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES)) {
              foundParams++;
              if (containsProducesLC = params.containsKey(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES_LC)) {
                  foundParams++;
              }
          }
          if (containsQ = params.containsKey("q")) {
              foundParams++;
          }
          if (containsQS = params.containsKey("qs")) {
              foundParams++;
          }

          // If there no parameters to remove, just return the existing chosen 
          if (foundParams > 0) {
              // If removing all parameters, check to see if it matches a static MediaType 
              if (foundParams == paramsSize) {
                  String key = chosen.getType().trim() + "/" + chosen.getSubtype().trim();
                  MediaType cached = mediaTypeCache.get(key);
                  if (cached != null) {
                      chosen = cached;
                  } else {
                      chosen = new MediaType(chosen.getType(), chosen.getSubtype());
                  }
              } else {
                  Map<String, String> map = new HashMap<String, String>(chosen.getParameters());
                  if (containsProduces) {
                      map.remove(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES);
                      if (containsProducesLC) {
                          map.remove(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES_LC);
                      }
                  }
                  if (containsQ) {
                      map.remove("q");
                  }
                  if (containsQS) {
                      map.remove("qs");
                  }
                  chosen = new MediaType(chosen.getType(), chosen.getSubtype(), map);
              }
          }
      }
```

Reply:
Thanks, your approach is a lot better than what I had.
I've made a couple of adjustments to shorten the simple path a bit, using the results of getParameters from the hasProduces evaluation early in the method, when available.